packages/nx-plugin/src/internal/execute-process.ts

```diff
@@ -0,0 +1,177 @@
+  stdout: string;
+  stderr: string;
+  code: number | null;
+  date: string;
+  duration: number;
+};
+
+/**
+ * Error class for process errors.
+ * Contains additional information about the process result.
+ * @category Error
+ * @public
+ * @class
+ * @extends Error
+ * @example
+ * const result = await executeProcess({})
+ * .catch((error) => {
+ *   if (error instanceof ProcessError) {
+ *   console.error(error.code);
+ *   console.error(error.stderr);
+ *   console.error(error.stdout);
+ *   }
+ * });
+ *
+ */
+export class ProcessError extends Error {
+  code: number | null;
+  stderr: string;
+  stdout: string;
+
+  constructor(result: ProcessResult) {
+    super(result.stderr);
+    this.code = result.code;
+    this.stderr = result.stderr;
+    this.stdout = result.stdout;
+  }
+}
+
+/**
+ * Process config object. Contains the command, args and observer.
+ * @param cfg - process config object with command, args and observer (optional)
+ * @category Types
+ * @public
+ * @property {string} command - The command to execute.
+ * @property {string[]} args - The arguments for the command.
+ * @property {ProcessObserver} observer - The observer for the process.
+ *
+ * @example
+ *
+ * // bash command
+ * const cfg = {
+ *   command: 'bash',
+ *   args: ['-c', 'echo "hello world"']
+ * };
+ *
+ * // node command
+ * const cfg = {
+ * command: 'node',
+ * args: ['--version']
+ * };
+ *
+ * // npx command
+ * const cfg = {
+ * command: 'npx',
+ * args: ['--version']
+ *
+ */
+export type ProcessConfig = {
+  command: string;
+  args?: string[];
+  cwd?: string;
+  observer?: ProcessObserver;
+  ignoreExitCode?: boolean;
+};
+
+/**
+ * Process observer object. Contains the onStdout, error and complete function.
+ * @category Types
+ * @public
+ * @property {function} onStdout - The onStdout function of the observer (optional).
+ * @property {function} onError - The error function of the observer (optional).
+ * @property {function} onComplete - The complete function of the observer (optional).
+ *
+ * @example
+ * const observer = {
+ *  onStdout: (stdout) => console.info(stdout)
+ *  }
+ */
+export type ProcessObserver = {
+  onStdout?: (stdout: string) => void;
+  onError?: (error: ProcessError) => void;
+  onComplete?: () => void;
+};
+
+/**
+ * Executes a process and returns a promise with the result as `ProcessResult`.
+ *
+ * @example
+ *
+ * // sync process execution
+ * const result = await executeProcess({
+ *  command: 'node',
+ *  args: ['--version']
+ * });
+ *
+ * console.info(result);
+ *
+ * // async process execution
+ * const result = await executeProcess({
+ *    command: 'node',
+ *    args: ['download-data.js'],
+ *    observer: {
+ *      onStdout: updateProgress,
+ *      error: handleError,
+ *      complete: cleanLogs,
+ *    }
+ * });
+ *
+ * console.info(result);
+ *
+ * @param cfg - see {@link ProcessConfig}
+ */
+export function executeProcess(cfg: ProcessConfig): Promise<ProcessResult> {
```

Comment:
Since the `executeProcess` function is already asynchronous, there's no disadvantage to having a dynamic import. This gets around the CJS vs ESM problem, so you can import the original implementation from `utils` without duplicating it.

```ts
export async function executeProcess(cfg: ProcessConfig): Promise<ProcessResult> {
  const { executeProcess } = await import('@code-pushup/utils');
  return executeProcess(cfg);
}
```

Reply:
Don't we have the problem that @poppins/cliui is not ESM ready?

Reply:
The problem is that the Nx Plugin has to CommonJS, which is the less compatible format. CJS can only import other CJS modules statically, while ESM can import both ESM and CJS. To import ESM (`utils`) into CJS (`nx-plugin`), you have to use a dynamic import. Because dynamic imports make the code async, it's incovienent for most of our utils. But for asynchronous functions like `executeProcess` the dynamic import doesn't add an extra inconvenience, since the usage is already async anyway.

As for `@poppins/cliui`, it's the other way around. The package is built only for the modern ESM format. That's a problem for CJS-only modules like Nx plugins.